### DISCUSSION TITLE
Range Error when using Bitwise NOT operator

### DISCUSSION THREAD

**Discussion by ProxyPlayerHD** [cc65/cc65#1797] [Q&A]
i was working on some macros to set and clear specific bits of NES control registers, when i came across this issue.
```
bitmask = %00000001
AND #~bitmask
```
causes a Range Error because `(-2 not in [0..255])`

This is really confusing because it implies that Bitwise NOT changes the signedness of it's input like the Negate Operator for some reason, and that Instructions do not accept signed operands.
and neither of which make a lot of sense to me...

what exactly is the reason why instruction operands cannot use negative numbers? the operand size is known while assembling so it couldn't be a potential size issue.
it would basically just change the operand range to `-128 to 255` (8-bit), `-32768 to 65535` (16-bit), etc for 24-bit and 32-bit. (exception being relative branches because they're a special case)
and also please have Bitwise NOT to keep the sign of the input...

EDIT:
```
bitmask = %00000001
AND #<(~bitmask)
```
this works, which miffed me a bit because that means it is a size issue when dealing with negative values. but why? can't it just be truncated down to the size of the operand instead of throwing an error?

**Comment by mrdudz** [comment]
It's not about negative numbers... the issue here is that in cases like this the assembler can not know how many bits are in a expression, so it has to assume the expression uses 24bits.

> can't it just be truncated down to the size of the operand instead of throwing an error?

No, this is exactly what we do NOT want - this would hide actual errors too.

**Reply by ProxyPlayerHD** [reply]
>It's not about negative numbers... the issue here is that in cases like this the assembler can not know how many bits are in a expression, so it has to assume the expression uses 24bits.

but as said, if the accepted range for a given operand size were simply extended to include negative numbers
(`-128 - 255` instead of ` 0 -255` for example), then wouldn't that avoid this issue regardless of the size of the expression compared to the operand it's supposed to be squeezed into?
`-2` is `-2` no matter how many bits are used to represent it, but something like `-129` for example would still throw a range error for an 8 bit operand.

>No, this is exactly what we do NOT want - this would hide actual errors too.

Ok yea i see why, i should've worded that differently, the idea was to "sign-shrink" numbers (ie check the bits beyond the sign bit, if they are all the same as the sign bit they can be cut off without changing the value of the number).
but i guess that would be functionally very similar or even identical to just extending the operand's accepted range.

(EDIT: meant to write this as a reply, oops)

**Comment by mrdudz** [comment]
> -2 is -2 no matter how many bits are used to represent it,

no - that is exactly the problem here. every value is represented in 24 bit, so %01 is really %00000000000000000000001, and the complement of that is %111111111111111111111110 (which is the -2 you see there). There is no way to automatically "shrink" this value.

**Reply by ProxyPlayerHD** [reply]
positive values do get automatically truncacted by the assembler, assuming that all cut bits that are 0
`%000000000000000011001100` -> `%11001100`
which makes sense as otherwise expressions wouldn't work at all without explicitly telling the assembler to cut off some bits.

so what exactly is keeping the assembler from doing the same with negative numbers? the values would always remain the same as it would just be shaving off 1's, exactly like when truncating the 0's off positive numbers.
Is it very difficult or even impossible to implement with the way the assembler works? or did just no-one want to bother implementing something like this yet? (maybe if it's low priority or something)

**Comment by mrdudz** [comment]
> so what exactly is keeping the assembler from doing the same with negative numbers? the values would always remain the same as it would just be shaving off 1's, exactly like when truncating the 0's off positive numbers.

ehrm, no. leading 0s have no value and can be ignored. however leading 1s do have a value.

To do what you want the assembler would have to track how many bits are in a value, which can really only work in some trivial cases - in all other cases it just cant be done - which is why the assembler does not do it.

**Reply by ProxyPlayerHD** [reply]
hmm, testing it i though the Assembler kept track of when values are seen as signed or unsigned, but it just uses regular signed 32-bit ints for everything. so $FFFFFFFE is the same as -2 in the eyes of the assembler.

oh well, guess this is actually impossible to implement without causing issues when dealing with full 32-bit expressions... or in this case when using logic operations. as AND, OR, XOR, and NOT are always applied to the full 32-bit expression.

**Comment by spiro-trikaliotis** [comment]
How do you know if the value $FFFFFE is a negative number (in which case truncating the upper bits is ok) or a positive number (in which case truncating the upper bits would be wrong!)?

EDIT: It seems @mrdudz and my answer were written at the same time. This was meant as an answer to @ProxyPlayerHD .

**Reply by ProxyPlayerHD** [reply]
the assembler uses 32-bit integers to keep track of expressions (even though mrdudz said they use 24-bit values?).
so $FFFFFE would always be seen as positive (16777214), while $FFFFFFFE would be seen as negative (-2).

but yea i can still see that being an issue when dealing with 32-bit addresses or values.

**Reply by mrdudz** [reply]
The assembler knows 8-, 16- and 24 bit values....

**Reply by ProxyPlayerHD** [reply]
It also knows 32-bit values. `.dword` exists for example.
The assembler accepts 32-bit values for expressions without any issues. `test_sym = 2147483647` or `test_sym = $DEADBEEF` work perfectly fine.
there is just no built-in way of directly accessing the highest byte unlike the 3 parts (<, ^, >). so you have to cheat a little bit to get that highest byte of the number.
example from an assembler generated list file:

```
000000r 1               test_sym = $02000000 + $07000000
000000r 1  00 00 00 09  .dword test_sym
000004r 1  A9 09        LDA #^(test_sym>>8)
```
or do you mean that there is no instruction that uses a 32-bit operand? which would be completely correct, but the assembler would technically be able to handle it.

**Comment by ghost** [comment]
I got hit by this issue today. Can we at least have casts?
e.g.
```
AND #.uint8(~bitmask)
```

**Comment by lgblgblgb** [comment]
@rodolfoschulz IIRC, I usually did this in this scenario:

`AND #.LOBYTE(~bitmask)`

**Reply by ghost** [reply]
oh yeah, that works too, thanks!
